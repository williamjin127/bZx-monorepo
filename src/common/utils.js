import { BigNumber } from "bignumber.js";

// const SCALE_FACTOR = BigNumber(1e18);

export const fromBigNumber = (bigNum, scale = 1) =>
  bigNum.div(scale).toNumber();

export const toBigNumber = (num, scale = 1) => {
  const bigNum = BigNumber(num);
  return bigNum.times(scale);
};
