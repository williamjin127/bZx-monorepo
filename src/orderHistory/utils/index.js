const ORDER_FIELD_COUNT = 14;
const SOMETHING = 64;

const remove0xPrefix = data => data.substr(2);
const getOrderObjArray = data =>
  data.match(new RegExp(`.{1,${ORDER_FIELD_COUNT * SOMETHING}}`, "g"));
const getOrder = data => data.match(new RegExp(`.{1,${SOMETHING}}`, "g"));

export const cleanData = raw => {
  const no0x = remove0xPrefix(raw);
  const objCount = no0x.length / SOMETHING / ORDER_FIELD_COUNT;
  if (objCount % 1 !== 0) throw new Error("Must be whole number of objects");

  const orderObjArray = getOrderObjArray(no0x);

  const hexRadix = 16;
  const orders = orderObjArray.map(orderObj => {
    const params = getOrder(orderObj);
    return {
      maker: `0x${params[0].substr(24)}`,
      loanTokenAddress: `0x${params[1].substr(24)}`,
      interestTokenAddress: `0x${params[2].substr(24)}`,
      collateralTokenAddress: `0x${params[3].substr(24)}`,
      feeRecipientAddress: `0x${params[4].substr(24)}`,
      oracleAddress: `0x${params[5].substr(24)}`,
      loanTokenAmount: parseInt(`0x${params[6]}`, hexRadix),
      interestAmount: parseInt(`0x${params[7]}`, hexRadix),
      initialMarginAmount: parseInt(`0x${params[8]}`, hexRadix),
      maintenanceMarginAmount: parseInt(`0x${params[9]}`, hexRadix),
      lenderRelayFee: parseInt(`0x${params[10]}`, hexRadix),
      traderRelayFee: parseInt(`0x${params[11]}`, hexRadix),
      expirationUnixTimestampSec: parseInt(`0x${params[12]}`, hexRadix),
      loanOrderHash: `0x${params[13]}`
    };
  });
  return orders;
};
