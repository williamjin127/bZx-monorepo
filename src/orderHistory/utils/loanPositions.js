import { pipe, map } from "ramda";
import * as Utils from "./index";

const NUM_LOAN_POS_FIELDS = 9;
const HEX_RADIX = 16;

const getLoanPosition = params => ({
  lender: `0x${params[0].substr(24)}`,
  trader: `0x${params[1].substr(24)}`,
  collateralTokenAddressFilled: `0x${params[2].substr(24)}`,
  positionTokenAddressFilled: `0x${params[3].substr(24)}`,
  loanTokenAmountFilled: parseInt(`0x${params[4]}`, HEX_RADIX),
  collateralTokenAmountFilled: parseInt(`0x${params[5]}`, HEX_RADIX),
  positionTokenAmountFilled: parseInt(`0x${params[6]}`, HEX_RADIX),
  loanStartUnixTimestampSec: parseInt(`0x${params[7]}`, HEX_RADIX),
  active: parseInt(`0x${params[8]}`, HEX_RADIX)
});

const checkProperObjCount = Utils.makeCheckProperObjCount(NUM_LOAN_POS_FIELDS);
const getOrderObjArray = Utils.makeGetOrderObjArray(NUM_LOAN_POS_FIELDS);

export const cleanData = raw =>
  pipe(
    Utils.remove0xPrefix,
    checkProperObjCount,
    getOrderObjArray,
    map(pipe(Utils.getOrderParams, getLoanPosition))
  )(raw);
