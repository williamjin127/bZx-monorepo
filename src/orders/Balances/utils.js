/* globals localStorage */

export const permaTokens = [`WETH_SM_ADDRESS_HERE`, `ZRX_SM_ADDRESS_HERE`];

export const getTrackedTokens = () => {
  const storedTokens = localStorage.getItem(`trackedTokens`);
  return storedTokens
    ? [...permaTokens, ...JSON.parse(storedTokens)]
    : permaTokens;
};

export const addTrackedToken = tokenAddress => {
  const currTokens = getTrackedTokens();
  const newTrackedTokens = currTokens.push(tokenAddress);
  return localStorage.setItem(`trackedTokens`, newTrackedTokens);
};

export const removeTrackedToken = tokenAddress => {
  const currTokens = getTrackedTokens();
  const newTrackedTokens = currTokens.filter(token => token !== tokenAddress);
  return localStorage.setItem(`trackedTokens`, newTrackedTokens);
};
