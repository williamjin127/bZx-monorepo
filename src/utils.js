import BigNumber from 'bignumber.js';
import BN from 'bn.js';
import ethABI from 'ethereumjs-abi';
import ethUtil from 'ethereumjs-util';
import _ from 'lodash';

import zeroExTypes from '0x.js/lib/src/types';

export const noop = () => {

};

export const bigNumberToBN = value => new BN(value.toString(), 10);

export const generatePseudoRandomSalt = () => {
  // BigNumber.random returns a pseudo-random number between 0 & 1
  // with a passed in number of decimal places.
  // Source: https://mikemcl.github.io/bignumber.js/#random
  const MAX_DIGITS_IN_UNSIGNED_256_INT = 78;
  const randomNumber = BigNumber.random(MAX_DIGITS_IN_UNSIGNED_256_INT);
  const factor = new BigNumber(10).pow(MAX_DIGITS_IN_UNSIGNED_256_INT - 1);
  const salt = randomNumber.times(factor).round();
  return salt;
};

export const getLendOrderHashHex = (order) => {
  const orderParams = [
    { value: order.b0x, type: zeroExTypes.SolidityTypes.Address },
    { value: order.maker, type: zeroExTypes.SolidityTypes.Address },
    { value: order.lendTokenAddress, type: zeroExTypes.SolidityTypes.Address },
    { value: order.interestTokenAddress, type: zeroExTypes.SolidityTypes.Address },
    { value: order.marginTokenAddress, type: zeroExTypes.SolidityTypes.Address },
    { value: order.feeRecipientAddress, type: zeroExTypes.SolidityTypes.Address },
    { value: bigNumberToBN(order.lendTokenAmount), type: zeroExTypes.SolidityTypes.Uint256 },
    { value: bigNumberToBN(order.interestAmount), type: zeroExTypes.SolidityTypes.Uint256 },
    { value: bigNumberToBN(order.initialMarginAmount), type: zeroExTypes.SolidityTypes.Uint256 },
    { value: bigNumberToBN(order.liquidationMarginAmount), type: zeroExTypes.SolidityTypes.Uint256 },
    { value: bigNumberToBN(order.lenderRelayFee), type: zeroExTypes.SolidityTypes.Uint256 },
    { value: bigNumberToBN(order.traderRelayFee), type: zeroExTypes.SolidityTypes.Uint256 },
    { value: bigNumberToBN(order.expirationUnixTimestampSec), type: zeroExTypes.SolidityTypes.Uint256 },
    { value: bigNumberToBN(order.salt), type: zeroExTypes.SolidityTypes.Uint256 },
  ];
  const types = _.map(orderParams, o => o.type);
  const values = _.map(orderParams, o => o.value);
  const hashBuff = ethABI.soliditySHA3(types, values);
  const orderHashHex = ethUtil.bufferToHex(hashBuff);
  return orderHashHex;
};
