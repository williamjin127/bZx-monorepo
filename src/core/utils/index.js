import { BigNumber } from "@0xproject/utils";
import { assert } from "@0xproject/assert";
import { constants } from "0x.js/lib/src/utils/constants";
import BN from "bn.js";
import Web3Utils from "web3-utils";
import * as Errors from "../constants/errors";
import { SchemaValidator } from "../../schemas/b0x_json_schemas";
import { ropsten as Contracts } from "../../contracts";
import * as Addresses from "../../addresses";

export const noop = () => {};

export const bigNumberToBN = value => new BN(value.toString(), 10);

export const generatePseudoRandomSalt = () => {
  // BigNumber.random returns a pseudo-random number between 0 & 1
  // with a passed in number of decimal places.
  // Source: https://mikemcl.github.io/bignumber.js/#random
  const randomNumber = BigNumber.random(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT
  );
  const factor = new BigNumber(10).pow(
    constants.MAX_DIGITS_IN_UNSIGNED_256_INT - 1
  );
  const salt = randomNumber.times(factor).round();
  return salt;
};

export const getLoanOrderHashHex = order => {
  const orderAddrs = [
    order.makerAddress,
    order.loanTokenAddress,
    order.interestTokenAddress,
    order.collateralTokenAddress,
    order.feeRecipientAddress,
    order.oracleAddress
  ];
  const orderUints = [
    bigNumberToBN(order.loanTokenAmount),
    bigNumberToBN(order.interestAmount),
    bigNumberToBN(order.initialMarginAmount),
    bigNumberToBN(order.maintenanceMarginAmount),
    bigNumberToBN(order.lenderRelayFee),
    bigNumberToBN(order.traderRelayFee),
    bigNumberToBN(order.expirationUnixTimestampSec),
    bigNumberToBN(order.makerRole),
    bigNumberToBN(order.salt)
  ];
  const orderHashHex = Web3Utils.soliditySha3(
    { t: "address", v: order.b0xAddress },
    { t: "address[6]", v: orderAddrs },
    { t: "uint256[9]", v: orderUints }
  );
  return orderHashHex;
};

export const doesContractExistAtAddress = async (web3, address) => {
  const code = await web3.eth.getCode(address);
  // Regex matches 0x0, 0x00, 0x in order to accommodate poorly implemented clients
  const codeIsEmpty = /^0x0{0,40}$/i.test(code);
  return !codeIsEmpty;
};

export const getContractInstance = async (web3, abi, address) => {
  assert.isETHAddressHex("address", address);
  const contractExists = await doesContractExistAtAddress(web3, address);
  if (!contractExists)
    throw new Error(`${Errors.ContractDoesNotExist} ${address}`);

  const contract = new web3.eth.Contract(abi, address);
  return contract;
};

export const doesConformToSchema = (variableName, value, schema) => {
  const schemaValidator = new SchemaValidator();
  const validationResult = schemaValidator.validate(value, schema);
  const hasValidationErrors = validationResult.errors.length > 0;
  const msg = `Expected ${variableName} to conform to schema ${
    schema.id
  }\nEncountered: ${JSON.stringify(
    value,
    null,
    "\t"
  )}\nValidation errors: ${validationResult.errors.join(", ")}`;
  assert.assert(!hasValidationErrors, msg);
};

export const isValidSignature = async (
  web3,
  { account, orderHash, signature }
) => {
  const b0xContract = await getContractInstance(
    web3,
    Contracts.B0x.abi,
    Addresses.getAddresses(web3.currentProvider).B0x
  );
  return b0xContract.methods
    .isValidSignature(account, orderHash, signature)
    .call();
};
