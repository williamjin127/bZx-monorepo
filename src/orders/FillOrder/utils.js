import styled from "styled-components";
import B0xJS from "b0x.js";  // eslint-disable-line
import { getTrackedTokens } from "../../common/trackedTokens";
import { toBigNumber } from "../../common/utils";

const TxHashLink = styled.a.attrs({
  target: `_blank`,
  rel: `noopener noreferrer`
})`
  font-family: monospace;
  display: block;
  text-overflow: ellipsis;
  overflow: auto;
}
`;

export const getOrderHash = order => B0xJS.getLoanOrderHashHex(order);

const checkAllowance = async (b0x, accounts, tokenAddress) => {
  const allowance = await b0x.getAllowance({
    tokenAddress,
    ownerAddress: accounts[0].toLowerCase()
  });
  return allowance.toNumber() !== 0;
};

const checkCoinsApproved = async (
  b0x,
  accounts,
  order,
  collateralTokenAddress
) => {
  const makerRole = order.makerRole === `0` ? `lender` : `trader`;
  if (makerRole === `lender`) {
    // check that user has approved collateralToken and interestToken
    const { interestTokenAddress } = order;
    const a = await checkAllowance(b0x, accounts, interestTokenAddress);
    const b = await checkAllowance(b0x, accounts, collateralTokenAddress);
    return a && b;
  }
  // check that user has approved loanToken
  const { loanTokenAddress } = order;
  const a = await checkAllowance(b0x, accounts, loanTokenAddress);
  return a;
};

export const validateFillOrder = async (
  order,
  fillOrderAmount,
  collateralTokenAddress,
  tokens,
  b0x,
  accounts
) => {
  const makerRole = order.makerRole === `0` ? `lender` : `trader`;
  const trackedTokens = getTrackedTokens(tokens);
  if (makerRole === `lender`) {
    // make sure the collateral token chosen is allowed
    const collateralToken = tokens.filter(
      t => t.address === collateralTokenAddress
    )[0];
    const notAllowed = {
      1: [],
      3: [`ZRX`, `B0X`],
      4: [],
      42: []
    };
    const collateralTokenNotAllowed = notAllowed[b0x.networkId].includes(
      collateralToken && collateralToken.symbol
    );
    if (collateralTokenNotAllowed) {
      alert(
        `The selected tokens are not yet supported for lending or collateral.`
      );
      return false;
    }

    // check that user has tracked collateralToken and interestToken
    const { interestTokenAddress } = order;
    if (
      !trackedTokens.includes(interestTokenAddress) ||
      !trackedTokens.includes(collateralTokenAddress)
    ) {
      alert(
        `Your interest token or collateral token is not tracked, please add it in the balances tab.`
      );
      return false;
    }
  } else {
    // check that user has tracked loanToken
    const { loanTokenAddress } = order;
    if (!trackedTokens.includes(loanTokenAddress)) {
      alert(
        `Your loan token is not tracked, please add it in the balances tab.`
      );
      return false;
    }
  }

  const coinsApproved = await checkCoinsApproved(
    b0x,
    accounts,
    order,
    collateralTokenAddress
  );
  if (!coinsApproved) {
    alert(
      `Some of your coins are not approved, please check the balances tab.`
    );
    return false;
  }
  console.log(`validateFillOrder`);
  console.log(order, fillOrderAmount, collateralTokenAddress);
  return true;
};
export const submitFillOrder = async (
  order,
  fillOrderAmount,
  collateralTokenAddress,
  b0x,
  accounts
) => {
  const txOpts = { from: accounts[0].toLowerCase() };
  const makerIsLender = order.makerRole === `0`;

  console.log(`order`, order);
  console.log(`txOpts`, txOpts);
  console.log(`makerIsLender`, makerIsLender);

  let receipt;
  if (makerIsLender) {
    // receipt = await b0x.takeLoanOrderAsTrader(
    //   order,
    //   collateralTokenAddress,
    //   toBigNumber(fillOrderAmount, 1e18),
    //   txOpts
    // );

    b0x
      .takeLoanOrderAsTrader(
        order,
        collateralTokenAddress,
        toBigNumber(fillOrderAmount, 1e18),
        txOpts
      )
      .once(`transactionHash`, hash => {
        alert(`Transaction submitted, transaction hash:`, {
          component: () => (
            <TxHashLink href={`${b0x.etherscanURL}tx/${hash}`}>
              {hash}
            </TxHashLink>
          )
        });
      })
      .on(`error`, error => {
        console.error(error.message);
      });
  } else {
    // receipt = await b0x.takeLoanOrderAsLender(order, txOpts);
    b0x
      .takeLoanOrderAsLender(order, txOpts)
      .once(`transactionHash`, hash => {
        alert(`Transaction submitted, transaction hash:`, {
          component: () => (
            <TxHashLink href={`${b0x.etherscanURL}tx/${hash}`}>
              {hash}
            </TxHashLink>
          )
        });
      })
      .on(`error`, error => {
        console.error(error.message);
      });
  }

  console.log(receipt);
  return true;
};
