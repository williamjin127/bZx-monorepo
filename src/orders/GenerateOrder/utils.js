import { getTokenInfo } from ".././../common/tokens";

export const compileObject = state => {
  const lendToken = getTokenInfo(state.lendTokenAddress);
  const interestToken = getTokenInfo(state.interestTokenAddress);
  const marginToken = getTokenInfo(state.marginTokenAddress);

  const { sendToRelayExchange } = state;

  return {
    makerAddress: `get_this_from_metamask`,
    b0xContract: `b0x_contract_address`,
    networkId: 1,

    // token data
    lendToken,
    interestToken,
    marginToken,

    // token amounts
    lendTokenAmount: state.lendTokenAmount,
    interestAmount: state.interestAmount,

    // margin amounts
    initialMarginAmount: state.initialMarginAmount,
    liquidationMarginAmount: state.liquidationMarginAmount,

    // expiration date/time
    expirationUnixTimestampSec: state.expirationDate.unix(),

    // relay/exchange settings
    feeRecipientAddress: sendToRelayExchange ? state.feeRecipientAddress : ``,
    lenderRelayFee: sendToRelayExchange ? state.lenderRelayFee : 0,
    traderRelayFee: sendToRelayExchange ? state.traderRelayFee : 0
  };
};

export const addSalt = obj => {
  // const salt = getLendOrderSalt();
  const salt = `my_salt_here`;
  return {
    ...obj,
    salt
  };
};

export const signOrder = obj => {
  // const signature = getSignature();
  const signature = {
    v: 27,
    r: `0x_temp`,
    s: `0x_temp`,
    hash: `0x_temp`
  };
  return {
    ...obj,
    signature
  };
};

// TODO - actually get the hash
export const getHash = () =>
  `0xa0443e64b09e95208424ec3bf7c1b543b841de766877a8b76e25d76b6b42b970`;
