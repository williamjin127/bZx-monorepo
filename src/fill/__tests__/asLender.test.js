import { pathOr } from "ramda";
import B0xJS from "../../core";
import b0xJS from "../../core/__tests__/setup";
import * as FillTestUtils from "./utils";
import { expectPromiEvent } from "../../core/__tests__/utils";

const { web3 } = b0xJS;
// Valid sig length, but last digit has been changed
const BAD_SIG =
  "0x056184af8d9bbf1734ddbff840e8be410193a99acab9add00512808250cb40f6423e669f24e65a8ee8af97e1a2abd90644177b39985438ecfee0dd2e7e44f77709";

describe("filling orders", () => {
  const { owner, lenders, traders } = FillTestUtils.getAccounts();
  const loanTokenAmount = web3.utils.toWei("251").toString();

  let promiEvent = null;

  beforeAll(async () => {
    const {
      loanTokens,
      collateralTokens,
      interestTokens
    } = FillTestUtils.initAllContractInstances();

    const transferAmount = web3.utils.toWei("500", "ether");
    await FillTestUtils.setupAll({ owner, lenders, traders, transferAmount });

    const takerAddress = lenders[1];
    const txOpts = {
      from: takerAddress,
      gas: 1000000,
      gasPrice: web3.utils.toWei("30", "gwei").toString()
    };

    const order = FillTestUtils.makeOrderAsTrader({
      web3,
      traders,
      loanTokens,
      interestTokens,
      collateralTokens,
      loanTokenAmount
    });

    const orderHashHex = B0xJS.getLoanOrderHashHex(order);
    const signature = await b0xJS.signOrderHashAsync(
      orderHashHex,
      order.makerAddress
    );

    promiEvent = b0xJS.takeLoanOrderAsLender({ ...order, signature }, txOpts);
  });

  describe("takeLoanOrderAsLender", async () => {
    test("should throw an error with an invalid signature", async () => {
      const {
        loanTokens,
        interestTokens,
        collateralTokens
      } = FillTestUtils.initAllContractInstances();

      const takerAddress = lenders[1];
      const txOpts = {
        from: takerAddress,
        gas: 1000000,
        gasPrice: web3.utils.toWei("30", "gwei").toString()
      };

      const order = FillTestUtils.makeOrderAsTrader({
        web3,
        traders,
        loanTokens,
        interestTokens,
        collateralTokens,
        loanTokenAmount
      });

      expect(() => {
        b0xJS.takeLoanOrderAsLender({ ...order, signature: BAD_SIG }, txOpts);
      }).toThrow();
      expect(() => {
        b0xJS.takeLoanOrderAsLender({ ...order, signature: BAD_SIG }, txOpts);
      }).toThrowErrorMatchingSnapshot();
    });

    test("should return total amount of loanToken borrowed", async () => {
      const receipt = await promiEvent;
      const loanTokenAmountFilledReturn = pathOr(
        null,
        ["events", "LogLoanTaken", "returnValues", "loanTokenAmountFilled"],
        receipt
      );

      expect(loanTokenAmountFilledReturn).toBe(loanTokenAmount);

      const debugLine = pathOr(null, ["events", "DebugLine"], receipt);
      expect(debugLine).toBe(null);
    });

    test("should return a web3 PromiEvent", async () => {
      expectPromiEvent(promiEvent);
    });
  });
});
