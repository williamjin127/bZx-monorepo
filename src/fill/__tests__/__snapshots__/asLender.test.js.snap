// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filling orders takeLoanOrderAsLender should throw an error with an invalid signature 1`] = `"Invalid signature v value"`;
